keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2861 @@
+enforcing compliance (and security). We intend to prioritize capabilities that
+make "fail closed" a safe mode of operation. As a concrete example, only
+allowing CEL expressions that pass compilation and type checking significantly
+reduces the opportunities for runtime errors.
+
+Also, making it possible (and convenient) to declare "zero trust" policies is
+important to compliance. By "zero trust", we mean policy rules that apply the
+principle of least privilege to newly created resources (e.g. namespaces) where
+the policy is initially set to the most restrictive state and can be made less
+restrictive via configuration.
+
+## Proposal
+
+Introduce a new `ValidatingAdmissionPolicy` kind to the
+admissionregistration.k8s.io group. (suggestions welcome on exact name to use
+for kind)
+
+At a high level, the API will support:
+
+- Request matching (similar to the match rules of admission webhooks, RBAC, priority & fairness and Audit)
+- CEL rule evaluation (similar to both [CRD Validation
+  Rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language)
+  but with access to the data in
+  [AdmissionRequest](https://github.com/kubernetes/kubernetes/blob/2ac6a4121f5b2a94acc88d62c07d8ed1cd34ed63/staging/src/k8s.io/api/admission/v1/types.go#L40))
+- Version conversion support (similar to admission webhook's MatchPolicy)
+- Access the old object (similar to [transition
+  rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language#transition-rules)
+  and oldObject in AdmissionRequest)
+- Configurability, as motivated above.
+
+There are also lots of additional capabilities (response message formatting,
+failure policies, type safety, advanced matching rules, ...) that will be discussed
+in detail further on in this proposal.
+
+We have divided this proposal into phases, all of which must be completed before
+this feature graduates to beta. Our goal is to size the phases so that each
+can be completed in a single Kubernetes release cycle.
+
+### Phase 1
+
+#### API Shape
+
+Before getting into all the individual fields and capabilities, let's look at the
+general "shape" of the API.
+
+This enhancement introduces a new `ValidatingAdmissionPolicy` kind.
+
+Each `ValidatingAdmissionPolicy` resource defines a admission control policy.
+The resource contains the CEL expressions to validate the admission policy and
+declares how the admission policy may be configured for use.
+
+For example:
+
+```yaml
+# Policy definition
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "replicalimit-policy.example.com"
+spec:
+  config:
+    group: rules.example.com
+    kind: ReplicaLimit
+    version: v1
+  match:
+    resourceRules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["deployments"]
```

Comment:
What happens if multiple resources are described here? Each config object applies to all of them?

Reply:
Yes, during admission, if the resource in the request is one of the resources in this array, it would be validated.

Reply:
So all of these have to typecheck against the CEL expression, right?

Reply:
I'm guessing the follow-up is "what about resources with different schemas?"

E.g. Someone wants to constrain workload resources so they check `CronJob`, `Deployment` and `Job`?

Reply:
Yes, exactly. And that will typecheck just fine if you use common fields, which for that set might just be metadata fields. But you have to do N different typechecks to know that.

Reply:
The "informational type checking" covers some of the wrenches that have been thrown at our type checking plans.

Some alternatives:

- We use GVK matching in policy definitions (only) and only allow two options: (a) Specify an exact Kind or (b) allow all kinds (useful for metadata only validations). Policy bindings still use GVR matching.
- Use GVR matching but only type check if all the match resources are of the same GVK (e.g. "*/scale")
- Type check against every GVK matched by the GVR match rules.
- Don't do any type checking.